[RUM-10500] [V3] Attributes encoding and validation

marco-saia-datadog opened the pull request:
### What does this PR do?

Starting from v3, we are changing the way attributes are handled to improve safety, consistency and compatibility with native SDKs, and to support more attribute types.

### What’s New

* Unsupported values (functions, symbols, non-finite numbers, etc.) are automatically dropped, with warnings, to prevent crashes in native SDKs.
* Common types such as `Date`, `Error`, and `Map` are properly serialized
* Nested objects are automatically flattened into dot syntax (`a.b.c = value`), matching the format expected by native SDKs and avoiding ambiguous outcomes.
* Custom encoders can be registered by users to support domain-specific types.
* Primitives and arrays at the root are wrapped under a `context` key to maintain schema consistency.

### Motivation

In React Native, it’s common for developers to pass complex objects as attributes (e.g. partial or full Redux/Zustand state, raw `Error` objects). These may contain unserializable values that would break serialization. While ideally users should only pass clean dictionaries of encodable values, in practice they expect the SDK to handle arbitrary objects to some extent.

This change ensures:

* Everything passed from JS to native is safe and JSON-serializable
* Flattening happens in JS, so users don’t need to handle dot syntax manually (which is not idiomatic in React Native), and we don't have to rely on the native SDKs either
* Ambiguities between nested objects and dot-syntax keys are avoided

### Example Transformations

* `"foo"` → `{ context: "foo" }`
* `{ user: { profile: { name: "Alice" } } }` → `{ "user.profile.name": "Alice" }`
* `new Error("Boom")` → `{ "context.name": "Error", "context.message": "Boom", "context.stack": "...", ... }`
* `new Map([["k1", 1], ["k2", "v2"]])` →

  ```json
  {
    "context": [
      { "key": "k1", "keyType": "string", "value": 1 },
      { "key": "k2", "keyType": "string", "value": "v2" }
    ]
  }
  ```
* Unsupported values like `Symbol("x")` or `() => {}` → `{}` (dropped with warning)

### Benchmarks

* **Flat/nested objects**: ~1M ops/sec
* **Deeply nested objects (20+ levels)**: ~233k ops/sec
* **Maps with mixed keys**: ~405k ops/sec
* **Arrays of objects**: ~28k–33k ops/sec (expectedly slower due to recursion)

marco-saia-datadog committed:
Attributes Safe Encoding

marco-saia-datadog committed:
Fixed existing tests

marco-saia-datadog committed:
Minor warning fixes in tests

sbarrio commented on packages/core/src/sdk/AttributesEncoding/__tests__/attributesEncoding.test.ts:
```diff
@@ -0,0 +1,266 @@
+        const result = encodeAttributes({
+            bad1: NaN,
+            bad2: Infinity,
+            good: 42
```
👌

sbarrio commented on packages/core/src/sdk/AttributesEncoding/defaultEncoders.tsx:
```diff
@@ -0,0 +1,185 @@
+    encode: (d: Date) => String(d)
+};
+/*
+            } else if ('componentStack' in error) {
```
Leftover?

sbarrio commented on packages/core/src/sdk/DdSdkInternal.ts:
```diff
@@ -0,0 +1,105 @@
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+import type {
```
Does this file need to be .tsx?, it's not using any JSX, right?

sbarrio commented on packages/core/src/sdk/AttributesEncoding/helpers.ts:
```diff
@@ -0,0 +1,134 @@
+/*
```
Does this file (as well as types.tsx  and utils.tsx) need to use .tsx?, they are not using any JSX syntax, right?

sbarrio reviewed (dismissed):
Left some minor comments. Good work!

marco-saia-datadog committed:
Changed .tsx to .ts + removed left-over comment

marco-saia-datadog commented:
Thanks @sbarrio 🙌 I have addressed your comments in a new commit. I still have to do some further manual testing before merging this. After I am done, I will squash the commits and merge this in v3 :)

marco-saia-datadog committed:
Limit encoded attributes to 128

marco-saia-datadog committed:
Add tests to ensure attributes are encoded by copy

marco-saia-datadog commented:
@sbarrio I made some changes to introduce the 128 attributes limit + additional testing for this new logic and for making sure encoding does not modify the original objects

sbarrio commented:
LGTM!

Just a heads up, due to the rename from DdSDK -> NativeDdSdk you will need to adapt all calls to the attributes API. These are the methods that will need adapting:

**Attributes API**

```
addAttribute
removeAttribute
addAttributes
removeAttributes
```

Please let me know if you want me to handle the merge or if you have any questions with it 🙂

sbarrio committed:
Missing attributes on FileBasedConfiguration and linting
